src/frontend/src/optimizer/share_parent_counter.rs

```diff
@@ -0,0 +1,49 @@
+// Copyright 2022 Singularity Data
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use std::collections::HashMap;
+
+use crate::optimizer::plan_node::{LogicalShare, PlanTreeNodeUnary};
+use crate::optimizer::plan_visitor::PlanVisitor;
+
+#[derive(Debug, Clone, Default)]
+pub struct ShareParentCounter {
+    // share id to parent number mapping.
```

Comment:
How about using a type alias or newtype for `ShareId`? Or can we directly use the `PlanNodeId` for this?

Reply:
Oops, actually it is plan node id.